### PR TITLE
feat: Allow P256 for EDV document encryption

### LINKED ISSUE #115: Allow P256 for EDV document encryption

### PR DESCRIPTION
**Pull request by DRK3**
Signed-off-by: Derek Trider <Derek.Trider@securekey.com>

closes #115

### CODE REVIEW

File: internal/cryptosetup/cryptosetup.go [hunk 1]
```diff
@@ -68,10 +64,58 @@ func PrepareMasterKeyReader(kmsSecretsStoreProvider ariesstorage.Provider) (*byt
...
+	}
+
+	ecPubKey := new(subtle.ECPublicKey)
+
+	err = unmarshal(buf.Bytes(), ecPubKey)
+	if err != nil {
+		return nil, err
+	}
+
+	jweEncrypter, err := newJWEEncrypt(jose.A256GCM, []subtle.ECPublicKey{*ecPubKey})
```
Comment (sandrask) [inline internal/cryptosetup/cryptosetup.go:107]: could have returned here

File: internal/cryptosetup/cryptosetup_test.go [hunk 2]
```diff
@@ -7,13 +7,20 @@ package cryptosetup
 
 import (
 	"encoding/base64"
+	"encoding/json"
 	"errors"
 	"testing"
 
+	"github.com/hyperledger/aries-framework-go/pkg/crypto/tinkcrypto/primitive/composite/ecdhes/subtle"
+	"github.com/hyperledger/aries-framework-go/pkg/doc/jose"
+
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup_test.go:16]: done in https://github.com/trustbloc/edge-service/pull/359

File: internal/cryptosetup/cryptosetup.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup.go:107]: done in https://github.com/trustbloc/edge-service/pull/359

File: internal/cryptosetup/cryptosetup.go [hunk 3]
```diff
@@ -68,10 +64,58 @@ func PrepareMasterKeyReader(kmsSecretsStoreProvider ariesstorage.Provider) (*byt
 	return masterKeyReader, nil
 }
 
+// PrepareJWECrypto prepares necessary JWE crypto data for edge-service operations
+func PrepareJWECrypto(keyManager kms.KeyManager,
+	storeProvider storage.Provider) (*jose.JWEEncrypt, *jose.JWEDecrypt, error) {
+	keyHandle, err := prepareKeyHandle(storeProvider, keyManager, ecdhesKeyIDDBKeyName, kms.ECDHES256AES256GCM)
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup.go:70]: done in https://github.com/trustbloc/edge-service/pull/359

File: internal/cryptosetup/cryptosetup_test.go [hunk 4]
```diff
@@ -43,55 +50,47 @@ func TestPrepareMasterKeyReader(t *testing.T) {
...
-		require.Empty(t, signingKey)
-		require.Nil(t, packer)
-	})
-	t.Run("Failure while creating key set when no existing signing key found", func(t *testing.T) {
-		mockStoreProvider := mockstore.NewMockStoreProvider()
-
-		mockKMS := legacykms.CloseableKMS{CreateKeyErr: errTest}
-
-		signingKey, packer, err := PrepareJWECrypto(&mockKMS, mockStoreProvider)
+		jweDecrypter, err := createJWEEncrypter(keyHandle, func(_ []byte, _ interface{}) error {
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup_test.go:71]: done in https://github.com/trustbloc/edge-service/pull/359

File: internal/cryptosetup/cryptosetup_test.go [hunk 5]
```diff
@@ -43,55 +50,47 @@ func TestPrepareMasterKeyReader(t *testing.T) {
...
+			return errTest
+		}, nil)
 		require.Equal(t, errTest, err)
-		require.Empty(t, signingKey)
-		require.Nil(t, packer)
+		require.Nil(t, jweDecrypter)
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup_test.go:75]: discussed, decided to leave as is for consistency

File: internal/cryptosetup/cryptosetup_test.go [hunk 6]
```diff
@@ -43,55 +50,47 @@ func TestPrepareMasterKeyReader(t *testing.T) {
...
-		mockStoreProvider := mockstore.NewMockStoreProvider()
-		mockStoreProvider.Store.ErrPut = errTest
-
-		mockKMS := legacykms.CloseableKMS{}
+	t.Run("Fail to create new JWE Encrypter", func(t *testing.T) {
+		keyHandle, err := keyset.NewHandle(ecdhes.ECDHES256KWAES256GCMKeyTemplate())
+		require.NoError(t, err)
 
-		signingKey, packer, err := PrepareJWECrypto(&mockKMS, mockStoreProvider)
+		jweDecrypter, err := createJWEEncrypter(keyHandle, json.Unmarshal,
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup_test.go:81]: done in https://github.com/trustbloc/edge-service/pull/359

File: internal/cryptosetup/cryptosetup_test.go [hunk 7]
```diff
@@ -43,55 +50,47 @@ func TestPrepareMasterKeyReader(t *testing.T) {
...
+			func(alg jose.EncAlg, keys []subtle.ECPublicKey) (*jose.JWEEncrypt, error) {
+				return nil, errTest
+			})
 		require.Equal(t, errTest, err)
-		require.Empty(t, signingKey)
-		require.Nil(t, packer)
+		require.Nil(t, jweDecrypter)
```
[the remark this answers is not in the document]
Comment (DRK3) [reply internal/cryptosetup/cryptosetup_test.go:86]: discussed, decided to leave as is for consistency

File: pkg/restapi/vc/operation/operations.go [hunk 8]
```diff
@@ -1749,14 +1750,19 @@ func (o *Operation) retrieveVC(profileName, docID, contextErrText string) ([]byt
...
-	decryptedEnvelope, err := o.packer.Unpack(document.JWE)
+	encryptedJWE, err := jose.Deserialize(string(document.JWE))
+	if err != nil {
+		return nil, err
+	}
+
+	decryptedDocBytes, err := o.jweDecrypter.Decrypt(encryptedJWE)
 	if err != nil {
-		return nil, fmt.Errorf("decrypted envelope unpacking failed while "+contextErrText+": %s", err)
+		return nil, fmt.Errorf("decrypting document failed while "+contextErrText+": %s", err)
```
[the remark this answers is not in the document]
Comment (DRK3) [reply pkg/restapi/vc/operation/operations.go:1760]: done in https://github.com/trustbloc/edge-service/pull/359

**Pull request merged**